File: 152-maximum-product-subarray/152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        
        int ans, currMax, currMin;
        ans = currMax = currMin = nums[0];
        
        for(int i=1; i<nums.size(); i++){
            int curr = nums[i];
            if(curr < 0) swap(currMax, currMin);
            
            currMin = min(curr, currMin*curr);
            currMax = max(curr, currMax*curr);
            
            cout<<"min is = "<<currMin<<endl;
            cout<<"max is = "<<currMax<<endl;
            
            ans = max(ans, currMax);
        }
        
        return ans;
    }
};

File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = max(nums)
        currMin, currMax = 1, 1
        
        for num in nums:
            if num == 0:
                currMin = currMax = 1
                continue
          
            temp = num*currMax
            currMax = max(num, temp, num*currMin)
            currMin = min(num, temp, num*currMin)
            
            res = max(res, currMax, currMin)
        return res